### DISCUSSION TITLE
Why can't unpacked arguments be used with TypeVarTuple parameters?

### DISCUSSION THREAD

**Discussion by gsingh93** [microsoft/pyright#5049] [Q&A] [answered]
This is the code I'm testing:
```python
from typing import Generic, Self, TypeVarTuple

S = TypeVarTuple("S")

class Shape(tuple, Generic[*S]):
    def __new__(cls, *args: *S) -> Self:
        return tuple.__new__(Shape, args)

    def __add__(self, other: Self, /) -> Self:
        x = (i + j for i, j in zip(self, other))
        return Shape(*x)
```

I'm getting this output for the `return Shape(*x)` line:
```
error: Unpacked argument cannot be used for TupleTypeVar parameter (reportGeneralTypeIssues)
```

Why is this the case? I couldn't find any explanation of this when searching online.

Sidenote: I'm guessing the error message should say "TypeVarTuple" instead of "TupleTypeVar"?

**Comment by erictraut** [comment, marked as answer]
Pyright doesn't support unpacked arguments for use with TypeVarTuple `*args` parameters unless the unpacked argument is itself a `TypeVarTuple`.

TypeVarTuples support (PEP 646) introduced _significant_ complexity and a huge number of edge cases to type checkers. Other type checkers like mypy have not yet implemented even basic support for the PEP two years after it was finalized. This is an edge case that I decided not to support in pyright's implementation to limit the complexity. Support for this case could be added in the future if we receive sufficient signal from pyright users that it is desired. For now, it will remain unsupported.

Thanks for pointing out the typo in the error message. It will be fixed in the next release.

**Reply by gsingh93** [reply]
Thanks, makes sense now.

>  Other type checkers like mypy have not yet implemented even basic support for the PEP two years after it was finalized

This is why I recently switched to pyright, thanks for all of the work on this 😄
